### PR TITLE
bf(S3C-4245): Enforce x-amz-bypass-governance-retention header and s3:BypassGovernanceRetention user policy

### PR DESCRIPTION
**Pull request by tmacro**
Second half of the object lock fix

### CODE REVIEW

File: lib/api/apiUtils/object/objectLockHelpers.js [hunk 1]
```diff
@@ -144,20 +144,24 @@ function isObjectLocked(bucket, objectMD, headers) {
...
     const existingDate = new Date(existingDateISO);
     const isExpired = existingDate < Date.now();
 
-    if (retentionInfo.mode === 'GOVERNANCE' && !isExpired) {
+    if (existingMode === 'GOVERNANCE' && !isExpired && !bypassGovernance) {
         return errors.AccessDenied;
     }
 
-    if (new Date(retentionInfo.date) < existingDate) {
+    if (existingMode === 'COMPLIANCE' && retentionInfo.mode === 'GOVERNANCE' && !isExpired) {
```
Comment (jonathan-gramain) [inline lib/api/apiUtils/object/objectLockHelpers.js]: Just for clarity, could benefit to group compliance checks in a block, or maybe even also grouping governance checks with a switch-like:

```javascript
if (existingMode === 'GOVERNANCE') {
    // governance checks
} else if (existingMode === 'COMPLIANCE') {
    // compliance checks
}
```

File: lib/api/objectPutRetention.js [hunk 2]
```diff
@@ -80,7 +82,51 @@ function objectPutRetention(authInfo, request, log, callback) {
...
+                        generalResource: bucketName,
+                        specificResource: { key: objectKey },
+                        requesterIp: ip,
+                        sslEnabled: request.connection.encrypted,
+                        apiMethod: 'bypassGovernanceRetention',
+                        awsService: 's3',
+                        locationConstraint: bucket.getLocationConstraint(),
+                        requesterInfo: authInfo,
+                        signatureVersion: authParams.params.data.authType,
+                        authType: authParams.params.data.signatureVersion,
```
Comment (rachedbenmustapha) [inline lib/api/objectPutRetention.js]: Swap here

Comment (jonathan-gramain) [reply lib/api/objectPutRetention.js]: makes me think that some specific tests to catch this particular typo issue could be useful (not sure what to test exactly, but I'm surprised that Vault would return a success if we mess with those input parameters)

Comment (tmacro) [reply lib/api/objectPutRetention.js]: After some testing, the value of these params doesn't matter as the values aren't used (though they still need to be present or vault will reject the request with `ValidationError`). The same error also exists in bucketPut and multiObjectDelete.

**Review by jonathan-gramain — APPROVED** [review]

**Review by rachedbenmustapha — APPROVED** [review]

**Pull request merged**